File: 0013-roman-to-integer/0013-roman-to-integer.js
/**
 * @param {string} s
 * @return {number}
 */
var romanToInt = function(s) {
    const special = new Map([
        ["CM", 900], ["CD", 400],
        ["XC", 90], ["XL", 40],
        ["IX", 9], ["IV", 4],
    ]);
    const single = new Map([
        ["M", 1000], ["D", 500], ["C", 100],
        ["L", 50], ["X", 10], ["V", 5], ["I", 1]
    ]);
    
    let total = 0;
    while (s != "") {
        if (s.length > 1 && special.has(s.slice(0, 2))) {
            total += special.get(s.slice(0, 2));
            s = s.slice(2);
        } else {
            total += single.get(s[0]);
            s = s.slice(1);
        }
    }
    
    return total;
};

File: 0013-roman-to-integer/0013-roman-to-integer.c
int convert_numeral(char c) {
    switch(c) {
        case 'I':
            return 1;
        case 'V':
            return 5;
        case 'X':
            return 10;
        case 'L':
            return 50;
        case 'C':
            return 100;
        case 'D':
            return 500;
        case 'M':
            return 1000;
    }
    return -1;
}

int romanToInt(char * s){
    int s_len = strlen(s);
    int total = 0;
    int last = convert_numeral(s[0]);
    
    for (int i = 0; i < s_len; i++) {
        int val = convert_numeral(s[i]);
        if (val > last) {
            total -= 2 * last;
        } 
        total += val;
        last = val;
    }
    
    return total;
}